### PR TITLE
sstables: Fix range of input stream in checksummed file data source

### LINKED ISSUE #21650: Compacting of sstables interrupted due to `sstables::malformed_sstable_exception`:  Checksummed chunk of size 0 at file offset
## Packages

Scylla version: `6.3.0~dev-20241119.733a4f94c7b2` with build-id `89a16d00ef82b3c15eb73e944afc60f15c478452`
Kernel Version: `6.8.0-1019-aws`

## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

`sstables::malformed_sstable_exception` while write load:

```
Nov 20 14:50:49.602174 perf-latency-nemesis-ubuntu-db-node-1621284f-3 scylla[6311]:  [shard 4:comp] compaction - [Compact keyspace1.standard1 d5155f10-a74e-11ef-88af-0a57f7f33336] Compacting
 [/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_154y_4i1c022tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756
120a74c11ef9d200a5bf7f33336/me-3gle_1515_11q0w22tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_152e_5wels22
tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_153o_2n5xc22tu7dhshimgm-big-Data.db:level=0:origin=memtable,
/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14zu_3r14022tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-9775612
0a74c11ef9d200a5bf7f33336/me-3gle_156a_4d3v422tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14yh_3uw0022tu
7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14uj_2j3bk22tu7dhshimgm-big-Data.db:level=0:origin=memtable,/v
ar/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14t9_5az0022tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a
74c11ef9d200a5bf7f33336/me-3gle_14vu_0rndc22tu7dhshimgm-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14x5_27ink22tu7d
hshimgm-big-Data.db:level=0:origin=memtable]
Nov 20 14:51:35.009059 perf-latency-nemesis-ubuntu-db-node-1621284f-3 scylla[6311]:  [shard 6:comp] compaction - [Compact keyspace1.standard1 af7f7bf0-a74e-11ef-b015-0a59f7f33336] Compacting of 2 sstables interrupted due to: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14vg_11alc2oebpia1otz5i-big-Data.db due to Checksummed chunk of size 0 at file offset 1415774208 failed checksum: expected=956923229, actual=0), at 0x6217fae 0x62185d0 0x62188d8 0x2574dda 0x257312e 0x60ecfe6
                                                                                       --------
                                                                                       --------
                                                                                       seastar::internal::coroutine_traits_base<sstables::compaction_result>::promise_type
                                                                                       --------
                                                                                       seastar::internal::coroutine_traits_base<std::optional<sstables::compaction_stats> >::promise_type
                                                                                       --------
                                                                                       seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state
Nov 20 14:51:35.010051 perf-latency-nemesis-ubuntu-db-node-1621284f-3 scylla[6311]:  [shard 6:comp] compaction_manager - Compaction task 0x6060043ad400 for table keyspace1.standard1 compaction_group=0 [0x606004d51a20]: failed: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14vg_11alc2oebpia1otz5i-big-Data.db due to Checksummed chunk of size 0 at file offset 1415774208 failed checksum: expected=956923229, actual=0). Will retry in 5 seconds

2024-11-20T14:51:35.270+00:00 perf-latency-nemesis-ubuntu-db-node-1621284f-3     !INFO | scylla[6311]:  [shard 6:comp] compaction - [Compact keyspace1.standard1 af7f7bf0-a74e-11ef-b015-0a59f7f33336] 
Compacting of 2 sstables interrupted due to: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14vg_11alc2oebpia1otz5i-big-Data.db due to Checksummed chunk of size 0 at file offset 1415774208 failed checksum: expected=956923229, actual=0), at 0x6217fae 0x62185d0 0x62188d8 0x2574dda 0x257312e 0x60ecfe6
```

```
Nov 20 14:49:29.917775 perf-latency-nemesis-ubuntu-db-node-1621284f-2 scylla[6305]:  [shard 3:comp] compaction - [Compact keyspace1.standard1 a59667c0-a74e-11ef-8621-d2c47dce7d22] Compacting
 [/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_1531_2kt1s21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756
120a74c11ef9d200a5bf7f33336/me-3gle_14yz_2fvkw21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_151o_4lgsg21
fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14rg_21qbk21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,
/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14xp_3xob421fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-9775612
0a74c11ef9d200a5bf7f33336/me-3gle_150b_4vjg021fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_154d_12sls21fj
sxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14sn_1mihc21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/v
ar/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14wg_2b5ts21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a
74c11ef9d200a5bf7f33336/me-3gle_14v6_17q2o21fjsxjwgrdia-big-Data.db:level=0:origin=memtable,/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14tw_1sq8w21fjsx
jwgrdia-big-Data.db:level=0:origin=memtable]
Nov 20 14:50:01.552497 perf-latency-nemesis-ubuntu-db-node-1621284f-2 scylla[6305]:  [shard 6:comp] compaction - [Compact keyspace1.standard1 82b851a0-a74e-11ef-8630-d2c07dce7d22] Compacting of 2 sstables interrupted due to: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14s6_5ih2821gpdjlapa90i-big-Data.db due to Checksummed chunk of size 0 at file offset 1049427968 failed checksum: expected=3787296288, actual=0), at 0x6217fae 0x62185d0 0x62188d8 0x2574dda 0x257312e 0x60ecfe6
                                                                                       --------
                                                                                       seastar::continuation<seastar::internal::promise_base_with_type<sstables::compaction_result>, seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda()#2}, seastar::future<void>::then_impl_nrvo<seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda()#2}, seastar::future<sstables::compaction_result> >(sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda(seastar::internal::promise_base_with_type<sstables::compaction_result>&&, seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, auto:1&&, (auto:2&&)...)::{lambda()#2}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
                                                                                       --------
                                                                                       seastar::continuation<seastar::internal::promise_base_with_type<sstables::compaction_result>, seastar::future<sstables::compaction_result>::finally_body<seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda()#3}, false>, seastar::future<sstables::compaction_result>::then_wrapped_nrvo<seastar::future<sstables::compaction_result>, seastar::future<sstables::compaction_result>::finally_body<seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda()#3}, false> >(seastar::future<sstables::compaction_result>::finally_body<seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0&&)::{lambda()#3}, false>&&)::{lambda(seastar::internal::promise_base_with_type<sstables::compaction_result>&&, seastar::future<sstables::compaction_result>::finally_body<seastar::async<sstables::compaction::run(std::unique_ptr<sstables::compaction, std::default_delete<sstables::compaction> >)::$_0>(seastar::thread_attributes, auto:1&&, (auto:2&&)...)::{lambda()#3}, false>&, seastar::future_state<sstables::compaction_result>&&)#1}, sstables::compaction_result>
                                                                                       --------
                                                                                       seastar::internal::coroutine_traits_base<sstables::compaction_result>::promise_type
                                                                                       --------
                                                                                       seastar::internal::coroutine_traits_base<std::optional<sstables::compaction_stats> >::promise_type
                                                                                       --------
                                                                                       seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state
Nov 20 14:50:01.553234 perf-latency-nemesis-ubuntu-db-node-1621284f-2 scylla[6305]:  [shard 6:comp] compaction_manager - Compaction task 0x606003ee3800 for table keyspace1.standard1 compaction_group=0 [0x606004b54a20]: failed: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14s6_5ih2821gpdjlapa90i-big-Data.db due to Checksummed chunk of size 0 at file offset 1049427968 failed checksum: expected=3787296288, actual=0). Will retry in 5 seconds
Nov 20 14:50:46.054386 perf-latency-nemesis-ubuntu-db-node-1621284f-2 scylla[6305]:  [shard 6:comp] compaction - [Compact keyspace1.standard1 82b89fc0-a74e-11ef-8630-d2c07dce7d22] Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_154v_2rg8w21gpdjlapa90i-big-Data.db:level=0]. 4GB to 4GB (~100% of original) in 134620ms = 30MB/s. ~3594880 total partitions merged to 3594803.

2024-11-20T14:50:02.042+00:00 perf-latency-nemesis-ubuntu-db-node-1621284f-2     !INFO | scylla[6305]:  [shard 6:comp] compaction - [Compact keyspace1.standard1 82b851a0-a74e-11ef-8630-d2c07dce7d22] 
Compacting of 2 sstables interrupted due to: sstables::malformed_sstable_exception (Failed to read partition from SSTable /var/lib/scylla/data/keyspace1/standard1-97756120a74c11ef9d200a5bf7f33336/me-3gle_14s6_5ih2821gpdjlapa90i-big-Data.db due to Checksummed chunk of size 0 at file offset 1049427968 failed checksum: expected=3787296288, actual=0), at 0x6217fae 0x62185d0 0x62188d8 0x2574dda 0x257312e 0x60ecfe6
```

Corrupted filed are collected and can be [downloaded](https://cloudius-jenkins-test.s3.us-east-1.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_153138/corrupfiles_1621284f.tar.gz)

## Impact

Compacting failed, data corruption

## How frequently does it reproduce?

Last successful run with Scylla version `6.3.0~dev-20241115.5bc03da0c4c1`:
https://argus.scylladb.com/tests/scylla-cluster-tests/378c37dd-51c6-4623-87f7-f5419a5d976f

Commits between `733a4f94c7b2` and `5bc03da0c4c1`:
```
git log --oneline --first-parent

733a4f94c7 Merge 'test/boost/view_schema_test.cc: Wait for views to build in test_view_update_generating_writetime' from Dawid Mędrek
7607f5e33e alternator: fix "/localnodes" to not return down nodes
980f6a48ab .github/scripts/auto-backport.py: validate backport candidate with `Fixes` prefix
165902b951 conf/scylla.yaml: update documentation for enable_tablets
36870feb29 Merge 'test: route S3 Proxy server messages through logger' from Kefu Chai
b14871ad3f Merge 'code cleanup: remove "sstring_view" and replace its usages by std::string_view' from Nadav Har'El
06d478793d Merge 'mutation: switch from boost ranges to std ranges' from Avi Kivity
34d7a4401d ./github/workflows/conflict_reminder.yaml: fix assignee object
572b005774 repair: implement tablet_repair_task_impl::release_resources
bef015da0d Revert ".github/scripts/auto-backport.py: validate backport candidate with `Fixes` prefix"
3a6c0a9b36 Merge 'compaction: Perform integrity checks on compacting SSTables' from Nikos Dragazis
f23800181a Merge 'Align Metric Family Descriptions' from Amnon Heiman
5bc03da0c4 tools/scylla-nodetool: rename estimated_row_count to estimated_partition_count
```

## Installation details

Cluster size: 3 nodes (i3en.2xlarge)

Scylla Nodes used in this run:
- perf-latency-nemesis-ubuntu-db-node-1621284f-3 (3.249.69.134 | 10.4.7.53) (shards: 7)
- perf-latency-nemesis-ubuntu-db-node-1621284f-2 (34.243.238.248 | 10.4.7.71) (shards: 7)
- perf-latency-nemesis-ubuntu-db-node-1621284f-1 (3.250.133.71 | 10.4.7.27) (shards: 7)

OS / Image: `ami-0d782b5935da36d59` (aws: undefined_region)

Test: `scylla-master-perf-regression-latency-650gb-with-nemesis`
Test id: `1621284f-c8ee-45d7-a758-2e9088b28e8f`
Test name: `scylla-master/perf-regression/scylla-master-perf-regression-latency-650gb-with-nemesis`
Test method: `performance_regression_test.PerformanceRegressionTest.test_latency_write_with_nemesis`
Test config file(s):

- [perf-regression-latency-650gb-with-nemesis.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/c124bfd47fdfcd23fab85a2fc1ddd011650b2e3f/test-cases/performance/perf-regression-latency-650gb-with-nemesis.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 1621284f-c8ee-45d7-a758-2e9088b28e8f`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=1621284f-c8ee-45d7-a758-2e9088b28e8f)
- Show all stored logs command: `$ hydra investigate show-logs 1621284f-c8ee-45d7-a758-2e9088b28e8f`

## Logs:
- **db-cluster-1621284f.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/db-cluster-1621284f.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/db-cluster-1621284f.tar.gz)
- **sct-runner-events-1621284f.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/sct-runner-events-1621284f.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/sct-runner-events-1621284f.tar.gz)
- **sct-1621284f.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/sct-1621284f.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/sct-1621284f.log.tar.gz)
- **loader-set-1621284f.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/loader-set-1621284f.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/loader-set-1621284f.tar.gz)
- **monitor-set-1621284f.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/monitor-set-1621284f.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_145419/monitor-set-1621284f.tar.gz)
- **corrupt files** - [https://cloudius-jenkins-test.s3.us-east-1.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_153138/corrupfiles_1621284f.tar.gz](https://cloudius-jenkins-test.s3.us-east-1.amazonaws.com/1621284f-c8ee-45d7-a758-2e9088b28e8f/20241120_153138/corrupfiles_1621284f.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/perf-regression/job/scylla-master-perf-regression-latency-650gb-with-nemesis/31/)
[Argus](https://argus.scylladb.com/test/4f088858-cbf2-4984-9d44-d319c207f521/runs?additionalRuns[]=1621284f-c8ee-45d7-a758-2e9088b28e8f)
</details>

### PR DESCRIPTION
**Pull request by ndragazis**
The checksummed file data source uses the chunk size to enforce that the reads from the underlying file input stream will be aligned at the chunk boundary. This is necessary so that we can validate the checksum of each chunk.

However, a mismatch in the numeric types caused a bug where the underlying file input stream would read a smaller portion of the data file than expected.

The bug is located in the following lines:

```
auto start = _beg_pos & ~(chunk_size - 1);
auto end = (_end_pos & ~(chunk_size - 1)) + chunk_size;
```

`_beg_pos` and `_end_pos` are `uint64_t`, whereas `chunk_size` is `uint32_t`. When executing the AND operation, the compiler converts the right operand from `uint32_t` to `uint64_t`. Since the integer is unsigned, the four most-significant bytes are filled with zeros, thus erroneously truncating the corresponding bytes of the position.

Fix the bug by explicitly converting the chunk size to `uint64_t` before any arithmetic operations. Also, replace the handwritten alignment implementations with the `align_up()` and `align_down()` helpers.

Finally, restrict the file end position to not exceed the file length. Since the last chunk can be smaller than the chunk size, it could happen that the end position exceeds the file length after the round-up. This is not a bug on its own since `make_file_input_stream()` can accept lengths that go beyond end-of-file, but still it makes the code more error prone and should be avoided.

Fixes #21650.

Needs to be backported to 6.2 that introduced the checksummed file data source. In that release, it only affects scrub/validate on large uncompressed SSTables (more than 4GiB).

### CODE REVIEW

**Comment by lersek** [comment]
I disagree with the patch, for three reasons:
1. The commit message talks about "zero extension". There is no such thing in C or C++. Instead, what happens is that the expression `(chunk_size - 1)` has type `uint32_t` (in practice, `unsigned int`), which is converted to `uint64_t` (in practice, `unsigned long`) according to the usual arithmetic conversions that the bitwise AND operator employs. And that conversion incurs no change to the value.
  In general, "sign extension" and "zero extension" are terms that are defined for some assembly instructions; they are meaningless in C and C++.
2. The proposed fix still detaches the required `uint64_t` from the actual expressions. If we insist on open-coded `~(chunk_size - 1)` expressions, then IMO it's better to write `~(uint64_t)(chunk_size - 1)` explicitly, every single time. Alternatively, we should use a "round up" helper function.
3. The patch does not fix the rounding error [pointed out](https://github.com/scylladb/scylladb/issues/21650#issuecomment-2493687802) by @michoecho. For rounding up the end, we should calculate `(_end_pos + (chunk_size - 1)) & ~(uint64_t)chunk_size`. When `end_pos` is already aligned (i.e., division by chunk_size yields a zero remainder), the addition does not affect the least significant chunk count bit, and is undone by the masking. Otherwise (remainder >=1), the chunk count is incremented by 1, and the sub-chunk-count bits are masked out. The addition wraps around if and only if the aligned-up value cannot be represented. All the more reason to use a helper function (or macro, or template...) for this calculation.

... In fact, Boost has `align_up`.

In summary: I suggest to eliminate the "zero extension" language from the commit message, and (instead of modifying the type of `chunk_size`) replacing *all* the upwards and downwards alignment bit-banging with Boost's `align_down` and `align_up`.

This is just my opinion of course; I'm not an authority on sstables.

**Comment by ndragazis** [comment]
Thanks for the review @lersek. A few comments:

> The commit message talks about "zero extension". There is no such thing in C or C++.

True. I was describing what happens at a lower level. How about this instead:
```
`_beg_pos` and `_end_pos` are `uint64_t`, whereas `chunk_size` is       
`uint32_t`. When executing the AND operation, the compiler promotes the 
right operand from `uint32_t` to `uint64_t`. Since the integer is       
unsigned, the four most-significant bytes are filled with zeros, thus    
erroneously truncating the corresponding bytes of the position.
```
> The proposed fix still detaches the required `uint64_t` from the actual expressions. If we insist on open-coded `~(chunk_size - 1)` expressions, then IMO it's better to write `~(uint64_t)(chunk_size - 1)` explicitly, every single time. Alternatively, we should use a "round up" helper function.

Actually, we do have some helpers in Seastar for this purpose. `align_up()` and `align_down()`. The cast is still required since they use the same template type parameter for the two arguments.

> The patch does not fix the rounding error https://github.com/scylladb/scylladb/issues/21650#issuecomment-2493687802 by @michoecho.

OK, Seastar's `align_up()` would fix that.  It's not a bug per se as @michoecho already pointed out. Also, exceeding the file length happens in all cases where the `_end_pos` falls within the last chunk. So, it's not an edge case either.  To cover all cases, we should truncate the aligned value at the file length.

**Comment by lersek** [comment]
@ndragazis

> ```
> `_beg_pos` and `_end_pos` are `uint64_t`, whereas `chunk_size` is       
> `uint32_t`. When executing the AND operation, the compiler promotes the 
> ```
I suggest "For the bitwise AND operator, the compiler *converts* ...".

(Emphasis not needed in the commit message; I'd just like to point out that what happens here is conversion, not promotion. Integer promotions exist, but they are a different thing. See `[conv.prom]` vs. `[expr.arith.conv]` in the standard. In particular, what takes effect here is paragraph (1.3.2) of `[expr.arith.conv]` -- *Otherwise, if T1 and T2 are both signed integer types or are both unsigned integer types, C is the type with greater rank.*)
> ```
> right operand from `uint32_t` to `uint64_t`. Since the integer is       
> unsigned, the four most-significant bytes are filled with zeros, thus    
> erroneously truncating the corresponding bytes of the position.
> ```
> 
> > The proposed fix still detaches the required `uint64_t` from the actual expressions. If we insist on open-coded `~(chunk_size - 1)` expressions, then IMO it's better to write `~(uint64_t)(chunk_size - 1)` explicitly, every single time. Alternatively, we should use a "round up" helper function.
> 
> Actually, we do have some helpers in Seastar for this purpose. `align_up()` and `align_down()`. The cast is still required since they use the same template type parameter for the two arguments.

Sounds good to me.

> 
> > The patch does not fix the rounding error [#21650 (comment)](https://github.com/scylladb/scylladb/issues/21650#issuecomment-2493687802) by @michoecho.
> 
> OK, Seastar's `align_up()` would fix that. It's not a bug per se as @michoecho already pointed out. Also, exceeding the file length happens in all cases where the `_end_pos` falls within the last chunk. So, it's not an edge case either. To cover all cases, we should truncate the aligned value at the file length.

Ah, indeed. Thanks for pointing that out!

**Comment by ndragazis** [comment]
>what happens here is conversion, not promotion. Integer promotions exist, but they are a different thing. See `[conv.prom]` vs. `[expr.arith.conv]` in the standard. In particular, what takes effect here is paragraph (1.3.2) of `[expr.arith.conv]` -- Otherwise, if T1 and T2 are both signed integer types or are both unsigned integer types, C is the type with greater rank.

I think I get it now, thanks. The section for the AND operator is explicit.
> The & operator groups left-to-right. The operands shall be of integral or unscoped enumeration type. The usual arithmetic conversions ([[expr.arith.conv]](https://eel.is/c++draft/expr.arith.conv)) are performed.

source: https://eel.is/c++draft/expr.bit.and

**Review by lersek — APPROVED** [review]
Indeed, `make_file_input_stream()` is documented as "the stream will stop at end-of-file even if `offset + len` is beyond end-of-file".

**Comment by avikivity** [comment]
@yaronkaikov why did docker start failing?

**Comment by denesb** [comment]
@ndragazis please annotate CI failures at https://github.com/scylladb/scylladb/pull/21665#issuecomment-2504744852

**Comment by ndragazis** [comment]
>Failed Tests (2/35897):
>
>    * [test_kill_coordinator_during_op.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13432/testReport/junit/topology_experimental_raft/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op.debug.3)
>    * [test_kill_coordinator_during_op.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13432/testReport/junit/topology_experimental_raft/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_debug_3_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op.debug.3)

Looks the same as #21114.

**Comment by lersek** [comment]
@ndragazis now that the patch is upstream, can we resolve #21650?

**Comment by ndragazis** [comment]
Hmm, I thought it would be closed automatically. It looks like we merge multi-commit PRs but rebase single-commit PRs, so the commit messages of the latter need to have a fixes/closes statement.

I'll close it manually, thanks.

**Comment by lersek** [comment]
Right, I suspected something like that: the PR description includes the proper `Fixes` line here, but for single-patch PRs, the PR description seems to be ignored. And the actual patch's commit message didn't include `Fixes`.
